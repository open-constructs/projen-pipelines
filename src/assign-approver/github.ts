import { GitHubProject, GithubWorkflow } from 'projen/lib/github';
import { JobPermission, JobPermissions } from 'projen/lib/github/workflows-model';
import { AssignApprover, AssignApproverOptions } from './base';

export interface GitHubAssignApproverOptions extends AssignApproverOptions {

  /**
   * runner tags to use to select runners
   *
   * @default ['ubuntu-latest']
   */
  readonly runnerTags?: string[];

}

export class GitHubAssignApprover extends AssignApprover {
  private readonly workflow: GithubWorkflow;
  private readonly options: GitHubAssignApproverOptions;

  constructor(scope: GitHubProject, options: GitHubAssignApproverOptions) {
    super(scope, options);
    this.options = options;

    // Initialize the deployment workflow on GitHub.
    this.workflow = scope.github!.addWorkflow('assign-approver');
    this.workflow.on({
      pullRequestTarget: {
        types: ['opened', 'ready_for_review'],
      },
    });

    this.setupWorkflow();
  }

  /**
   * Get the required permissions for the GitHub workflow
   */
  public renderPermissions(): JobPermissions {
    return {
      pullRequests: JobPermission.WRITE,
    };
  }

  protected setupWorkflow(): void {
    const runnerTags = this.options.runnerTags ?? ['ubuntu-latest'];

    const approverMappingScript = this.generateApproverMappingScript();

    this.workflow.addJob('assign-approver', {
      runsOn: runnerTags,
      permissions: this.renderPermissions(),
      steps: [
        {
          name: 'Assign approver based on author',
          uses: 'actions/github-script@v7',
          with: {
            script: approverMappingScript,
          },
        },
      ],
    });
  }

  protected generateApproverMappingScript(): string {
    const mappingEntries = this.baseOptions.approverMapping
      .map(mapping => `  '${mapping.author}': [${mapping.approvers.map(a => `'${a}'`).join(', ')}]`)
      .join(',\n');

    const defaultApprovers = this.baseOptions.defaultApprovers
      .map(a => `'${a}'`)
      .join(', ');

    return `const author = context.payload.pull_request.user.login;

// Define approver mapping
const approverMapping = {
${mappingEntries}${mappingEntries.length > 0 ? ',' : ''}
  'default': [${defaultApprovers}] // Default approver(s) if author not in mapping
};

// Get approvers for the PR author
const approvers = approverMapping[author] || approverMapping['default'];

// Filter out the author from approvers list (can't approve own PR)
const filteredApprovers = approvers.filter(approver => approver !== author);

if (filteredApprovers.length > 0) {
  // Request reviews from the approvers
  await github.rest.pulls.requestReviewers({
    owner: context.repo.owner,
    repo: context.repo.repo,
    pull_number: context.issue.number,
    reviewers: filteredApprovers
  });
  
  console.log(\`Assigned reviewers: \${filteredApprovers.join(', ')}\`);
} else {
  console.log('No eligible reviewers found for this PR author');
}`;
  }
}
