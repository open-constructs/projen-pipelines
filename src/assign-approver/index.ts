export * from './base';
export * from './github';