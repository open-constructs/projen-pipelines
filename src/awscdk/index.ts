export * from './base';
export * from './github';
export * from './gitlab';
export * from './codecatalyst';
export * from './bash';
