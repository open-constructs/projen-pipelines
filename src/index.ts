export * from './engine';
export * from './steps';
export * from './awscdk';