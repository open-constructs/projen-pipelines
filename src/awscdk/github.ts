import { awscdk } from 'projen';
import { GithubWorkflow } from 'projen/lib/github';
import { JobPermission, JobStep } from 'projen/lib/github/workflows-model';
import { CDKPipeline, CDKPipelineOptions, DeploymentStage } from './base';
import { PipelineEngine } from '../engine';

const DEFAULT_RUNNER_TAGS = ['ubuntu-latest'];

/**
 * Configuration interface for GitHub-specific IAM roles used in the CDK pipeline.
 */
export interface GithubIamRoleConfig {

  /** Default IAM role ARN used if no specific role is provided. */
  readonly default?: string;
  /** IAM role ARN for the synthesis step. */
  readonly synth?: string;
  /** IAM role ARN for the asset publishing step. */
  readonly assetPublishing?: string;
  /** IAM role ARNs for different deployment stages. */
  readonly deployment?: { [stage: string]: string };
}

/**
 * Extension of the base CDKPipeline options including specific configurations for GitHub.
 */
export interface GithubCDKPipelineOptions extends CDKPipelineOptions {

  /** IAM config for GitHub Actions */
  readonly iamRoleArns: GithubIamRoleConfig;

  /**
   * runner tags to use to select runners
   *
   * @default ['ubuntu-latest']
   */
  readonly runnerTags?: string[];
}


/**
 * Implements a CDK Pipeline configured specifically for GitHub workflows.
 */
export class GithubCDKPipeline extends CDKPipeline {

  /** Indicates if versioned artifacts are needed based on manual approval requirements. */
  public readonly needsVersionedArtifacts: boolean;

  /** The GitHub workflow associated with the pipeline. */
  private deploymentWorkflow: GithubWorkflow;
  /** List of deployment stages for the pipeline. */
  private deploymentStages: string[] = [];

  /**
   * Constructs a new GithubCDKPipeline instance.
   * @param app - The CDK app associated with this pipeline.
   * @param options - Configuration options for the pipeline.
   */
  constructor(app: awscdk.AwsCdkTypeScriptApp, private options: GithubCDKPipelineOptions) {
    super(app, options);

    // Initialize the deployment workflow on GitHub.
    this.deploymentWorkflow = this.app.github!.addWorkflow('deploy');
    this.deploymentWorkflow.on({
      push: {
        branches: [this.branchName],
      },
      workflowDispatch: {},
    });

    // Determine if versioned artifacts are necessary.
    this.needsVersionedArtifacts = this.options.stages.find(s => s.manualApproval === true) !== undefined;

    // Create jobs for synthesizing, asset uploading, and deployment.
    this.createSynth();

    this.createAssetUpload();

    for (const stage of options.stages) {
      this.createDeployment(stage);
    }
  }

  /** the type of engine this implementation of CDKPipeline is for */
  public engineType(): PipelineEngine {
    return PipelineEngine.GITHUB;
  }

  /**
   * Creates a synthesis job for the pipeline using GitHub Actions.
   */
  private createSynth(): void {
    const steps: JobStep[] = [{
      name: 'Checkout',
      uses: 'actions/checkout@v4',
    }];

    if (this.options.iamRoleArns?.synth) {
      steps.push({
        name: 'AWS Credentials',
        uses: 'aws-actions/configure-aws-credentials@master',
        with: {
          'role-to-assume': this.options.iamRoleArns.synth,
          'role-session-name': 'GitHubAction',
          'aws-region': 'us-east-1',
        },
      });
    }
    const preInstallSteps = (this.options.preInstallSteps ?? []).map(s => s.toGithub());
    const preSynthSteps = (this.options.preSynthSteps ?? []).map(s => s.toGithub());
    const postSynthSteps = (this.options.postSynthSteps ?? []).map(s => s.toGithub());

    steps.push(...preInstallSteps.flatMap(s => s.steps));
    steps.push(...this.renderInstallCommands().map(cmd => ({
      run: cmd,
    })));

    steps.push(...preSynthSteps.flatMap(s => s.steps));
    steps.push(...this.renderSynthCommands().map(cmd => ({
      run: cmd,
    })));
    steps.push(...postSynthSteps.flatMap(s => s.steps));

    steps.push({
      uses: 'actions/upload-artifact@v4',
      with: {
        name: 'cloud-assembly',
        path: `${this.app.cdkConfig.cdkout}/`,
      },
    });

    this.deploymentWorkflow.addJob('synth', {
      name: 'Synth CDK application',
      runsOn: this.options.runnerTags ?? DEFAULT_RUNNER_TAGS,
      env: {
        CI: 'true',
      },
      needs: [...preInstallSteps.flatMap(s => s.needs), ...preSynthSteps.flatMap(s => s.needs), ...postSynthSteps.flatMap(s => s.needs)],
      permissions: { idToken: JobPermission.WRITE, contents: JobPermission.READ },
      steps,
    });
  }

  /**
   * Creates a job to upload assets to AWS as part of the pipeline.
   */
  public createAssetUpload(): void {
    const preInstallSteps = (this.options.preInstallSteps ?? []).map(s => s.toGithub());

    this.deploymentWorkflow.addJob('assetUpload', {
      name: 'Publish assets to AWS',
      needs: ['synth', ...preInstallSteps.flatMap(s => s.needs)],
      runsOn: this.options.runnerTags ?? DEFAULT_RUNNER_TAGS,
      env: {
        CI: 'true',
      },
      permissions: { idToken: JobPermission.WRITE, contents: this.needsVersionedArtifacts ? JobPermission.WRITE : JobPermission.READ },
      steps: [{
        name: 'Checkout',
        uses: 'actions/checkout@v4',
        with: {
          'fetch-depth': 0,
        },
      }, {
        name: 'Setup GIT identity',
        run: 'git config --global user.name "github-actions" && git config --global user.email "github-actions@github.com"',
      }, {
        name: 'AWS Credentials',
        uses: 'aws-actions/configure-aws-credentials@master',
        with: {
          'role-to-assume': this.options.iamRoleArns?.assetPublishing ?? this.options.iamRoleArns?.default,
          'role-session-name': 'GitHubAction',
          'aws-region': 'us-east-1',
        },
      }, {
        uses: 'actions/download-artifact@v4',
        with: {
          name: 'cloud-assembly',
          path: `${this.app.cdkConfig.cdkout}/`,
        },
      },
      ...preInstallSteps.flatMap(s => s.steps),
      ...this.renderInstallCommands().map(cmd => ({
        run: cmd,
      })),
      ...this.getAssetUploadCommands(this.needsVersionedArtifacts).map(cmd => ({
        run: cmd,
      }))],
    });
  }

  /**
   * Creates a job to deploy the CDK application to AWS.
   * @param stage - The deployment stage to create.
   */
  public createDeployment(stage: DeploymentStage): void {
    const preInstallSteps = (this.options.preInstallSteps ?? []).map(s => s.toGithub());

    if (stage.manualApproval === true) {
      // Create new workflow for deployment
      const stageWorkflow = this.app.github!.addWorkflow(`release-${stage.name}`);
      stageWorkflow.on({
        workflowDispatch: {
          inputs: {
            version: {
              description: 'Package version',
              required: true,
            },
          },
        },
      });
      stageWorkflow.addJob('deploy', {
        name: `Release stage ${stage.name} to AWS`,
        needs: preInstallSteps.flatMap(s => s.needs),
        runsOn: this.options.runnerTags ?? DEFAULT_RUNNER_TAGS,
        env: {
          CI: 'true',
        },
        permissions: { idToken: JobPermission.WRITE, contents: JobPermission.READ },
        steps: [{
          name: 'Checkout',
          uses: 'actions/checkout@v4',
        }, {
          name: 'AWS Credentials',
          uses: 'aws-actions/configure-aws-credentials@master',
          with: {
            'role-to-assume': this.options.iamRoleArns?.deployment?.[stage.name] ?? this.options.iamRoleArns?.default,
            'role-session-name': 'GitHubAction',
            'aws-region': stage.env.region,
          },
        },
        ...preInstallSteps.flatMap(s => s.steps),
        ...this.renderInstallCommands().map(cmd => ({
          run: cmd,
        })),
        ...this.renderInstallPackageCommands(`${this.options.pkgNamespace}/${this.app.name}@\${{github.event.inputs.version}}`).map(cmd => ({
          run: cmd,
        })),
        {
          run: `mv ./node_modules/${this.options.pkgNamespace}/${this.app.name} ${this.app.cdkConfig.cdkout}`,
        },
        ...this.renderDeployCommands(stage.name).map(cmd => ({
          run: cmd,
        })),
        {
          uses: 'actions/upload-artifact@v3',
          with: {
            name: `cdk-outputs-${stage.name}`,
            path: `cdk-outputs-${stage.name}.json`,
          },
        }],
      });

    } else {
      // Add deployment to CI/CD workflow
      this.deploymentWorkflow.addJob(`deploy-${stage.name}`, {
        name: `Deploy stage ${stage.name} to AWS`,
        needs: ['assetUpload', ...preInstallSteps.flatMap(s => s.needs), ...(this.deploymentStages.length > 0 ? [`deploy-${this.deploymentStages.at(-1)!}`] : [])],
        runsOn: this.options.runnerTags ?? DEFAULT_RUNNER_TAGS,
        env: {
          CI: 'true',
        },
        permissions: { idToken: JobPermission.WRITE, contents: JobPermission.READ },
        steps: [{
          name: 'Checkout',
          uses: 'actions/checkout@v4',
        }, {
          name: 'AWS Credentials',
          uses: 'aws-actions/configure-aws-credentials@master',
          with: {
            'role-to-assume': this.options.iamRoleArns?.deployment?.[stage.name] ?? this.options.iamRoleArns?.default,
            'role-session-name': 'GitHubAction',
            'aws-region': stage.env.region,
          },
        }, {
          uses: 'actions/download-artifact@v4',
          with: {
            name: 'cloud-assembly',
            path: `${this.app.cdkConfig.cdkout}/`,
          },
        },
        ...preInstallSteps.flatMap(s => s.steps),
        ...this.renderInstallCommands().map(cmd => ({
          run: cmd,
        })),
        ...this.renderDeployCommands(stage.name).map(cmd => ({
          run: cmd,
        })),
        {
          uses: 'actions/upload-artifact@v3',
          with: {
            name: `cdk-outputs-${stage.name}`,
            path: `cdk-outputs-${stage.name}.json`,
          },
        }],
      });
      this.deploymentStages.push(stage.name);
    }
  }
}
