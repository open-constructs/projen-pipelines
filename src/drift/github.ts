import { Project } from 'projen';
import { GitHubProject, GithubWorkflow } from 'projen/lib/github';
import { JobPermission } from 'projen/lib/github/workflows-model';
import { DriftDetectionWorkflow, DriftDetectionWorkflowOptions, DriftDetectionStageOptions } from './base';
import { DriftDetectionStep } from './step';

export interface GitHubDriftDetectionWorkflowOptions extends DriftDetectionWorkflowOptions {
  /**
   * Additional permissions for GitHub workflow
   */
  readonly permissions?: Record<string, string>;

  /**
   * Whether to create issues on drift detection
   * @default false
   */
  readonly createIssues?: boolean;
}

export class GitHubDriftDetectionWorkflow extends DriftDetectionWorkflow {
  private readonly permissions?: Record<string, string>;
  private readonly createIssues: boolean;
  private readonly workflow: GithubWorkflow;

  constructor(project: Project, options: GitHubDriftDetectionWorkflowOptions) {
    super(project, options);
    this.permissions = options.permissions;
    this.createIssues = options.createIssues ?? false;

    this.workflow = (this.project as GitHubProject).github!.addWorkflow('drift-detection');
    this.workflow.on({
      schedule: [{
        cron: this.schedule,
      }],
      workflowDispatch: {
        inputs: {
          stage: {
            description: 'Stage to check for drift (leave empty for all)',
            required: false,
            type: 'choice',
            options: this.stages.map(s => s.name),
          },
        },
      },
    });

    // Add job for each stage
    for (const stage of this.stages) {
      const jobId = `drift-${stage.name}`.toLowerCase().replace(/[^a-z0-9-]/g, '-');

      const driftStep = new DriftDetectionStep(this.project, stage).toGithub();

      this.workflow.addJob(jobId, {
        name: `Drift Detection - ${stage.name}`,
        runsOn: ['ubuntu-latest'],
        if: `\${{ github.event_name == 'schedule' || github.event.inputs.stage == '' || github.event.inputs.stage == '${stage.name}' }}`,
        env: driftStep.env,
        permissions: {
          contents: JobPermission.READ,
          ...(driftStep.permissions ?? {}),
          ...(this.createIssues ? { issues: JobPermission.WRITE } : {}),
          ...this.permissions,
        },
        steps: [
          {
            name: 'Checkout',
            uses: 'actions/checkout@v4',
          },
          {
            name: 'Setup Node.js',
            uses: 'actions/setup-node@v4',
            with: {
              'node-version': '20',
            },
          },
          {
            name: 'Install dependencies',
            run: 'npm ci',
          },
          ...driftStep.steps,
          {
            name: 'Upload results',
            uses: 'actions/upload-artifact@v4',
            with: {
              name: `drift-results-${stage.name}`,
              path: `drift-results-${stage.name}.json`,
            },
          },
          ...(this.createIssues ? [{
            name: 'Create Issue on Drift',
            if: 'steps.drift.outcome == \'failure\' && github.event_name == \'schedule\'',
            uses: 'actions/github-script@v7',
            with: {
              script: this.generateIssueCreationScript(stage),
            },
          }] : []),
        ],
      });
    }

    // Add summary job
    if (this.stages.length > 0) {
      this.workflow.addJob('drift-summary', {
        name: 'Drift Detection Summary',
        runsOn: ['ubuntu-latest'],
        permissions: {
          contents: JobPermission.READ,
        },
        needs: this.stages.map(stage => `drift-${stage.name}`),
        steps: [
          {
            name: 'Download all artifacts',
            uses: 'actions/download-artifact@v4',
            with: {
              path: 'drift-results',
            },
          },
          {
            name: 'Generate summary',
            run: this.generateSummaryScript(),
          },
        ],
      });
    }
  }

  private generateIssueCreationScript(stage: DriftDetectionStageOptions): string {
    return `
const fs = require('fs');
const resultsFile = 'drift-results-${stage.name}.json';

if (!fs.existsSync(resultsFile)) {
  console.log('No results file found');
  return;
}

const results = JSON.parse(fs.readFileSync(resultsFile, 'utf8'));
const driftedStacks = results.filter(r => r.driftStatus === 'DRIFTED');

if (driftedStacks.length === 0) {
  console.log('No drift detected');
  return;
}

const title = 'Drift Detected in ${stage.name}';
const body = \`## Drift Detection Report

**Stage:** ${stage.name}
**Region:** ${stage.region}
**Time:** \${new Date().toISOString()}

### Summary
- Total stacks checked: \${results.length}
- Drifted stacks: \${driftedStacks.length}

### Drifted Stacks
\${driftedStacks.map(stack => {
  const resources = stack.driftedResources || [];
  return \`#### \${stack.stackName}
- Drifted resources: \${resources.length}
\${resources.map(r => \`  - \${r.logicalResourceId} (\${r.resourceType})\`).join('\\n')}
\`;
}).join('\\n')}

### Action Required
Please review the drifted resources and either:
1. Update the infrastructure code to match the actual state
2. Restore the resources to match the expected state

[View workflow run](\${context.serverUrl}/\${context.repo.owner}/\${context.repo.repo}/actions/runs/\${context.runId})
\`;

// Check if issue already exists
const issues = await github.rest.issues.listForRepo({
  owner: context.repo.owner,
  repo: context.repo.repo,
  state: 'open',
  labels: ['drift-detection', '${stage.name}'],
});

if (issues.data.length === 0) {
  await github.rest.issues.create({
    owner: context.repo.owner,
    repo: context.repo.repo,
    title,
    body,
    labels: ['drift-detection', '${stage.name}'],
  });
} else {
  // Update existing issue
  const issue = issues.data[0];
  await github.rest.issues.createComment({
    owner: context.repo.owner,
    repo: context.repo.repo,
    issue_number: issue.number,
    body: body,
  });
}
`;
  }

  private generateSummaryScript(): string {
    return `
#!/bin/bash
echo "## Drift Detection Summary" >> $GITHUB_STEP_SUMMARY
echo "" >> $GITHUB_STEP_SUMMARY

total_stacks=0
total_drifted=0
total_errors=0

for file in drift-results-*.json; do
  if [[ -f "$file" ]]; then
    stage=$(basename $(dirname "$file"))
    echo "### Stage: $stage" >> $GITHUB_STEP_SUMMARY
    
    # Parse JSON and generate summary
    jq -r '
      . as $results |
      "- Total stacks: " + ($results | length | tostring) + "\\n" +
      "- Drifted: " + ([$results[] | select(.driftStatus == "DRIFTED")] | length | tostring) + "\\n" +
      "- Errors: " + ([$results[] | select(.error)] | length | tostring) + "\\n" +
      ([$results[] | select(.driftStatus == "DRIFTED")] | 
        if length > 0 then
          "\\n**Drifted stacks:**\\n" + 
          (map("  - " + .stackName + " (" + ((.driftedResources // []) | length | tostring) + " resources)") | join("\\n"))
        else "" end)
    ' "$file" >> $GITHUB_STEP_SUMMARY
    
    echo "" >> $GITHUB_STEP_SUMMARY
    
    # Count totals
    total_stacks=$((total_stacks + $(jq 'length' "$file")))
    total_drifted=$((total_drifted + $(jq '[.[] | select(.driftStatus == "DRIFTED")] | length' "$file")))
    total_errors=$((total_errors + $(jq '[.[] | select(.error)] | length' "$file")))
  fi
done

echo "### Overall Summary" >> $GITHUB_STEP_SUMMARY
echo "- Total stacks checked: $total_stacks" >> $GITHUB_STEP_SUMMARY
echo "- Total drifted stacks: $total_drifted" >> $GITHUB_STEP_SUMMARY
echo "- Total errors: $total_errors" >> $GITHUB_STEP_SUMMARY

if [[ $total_drifted -gt 0 ]]; then
  echo "" >> $GITHUB_STEP_SUMMARY
  echo "⚠️ **Action required:** Drift detected in $total_drifted stacks" >> $GITHUB_STEP_SUMMARY
fi
`;
  }

}