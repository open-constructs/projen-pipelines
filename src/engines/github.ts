import { JobPermission, JobPermissions } from 'projen/lib/github/workflows-model';

/**
 * Merge multiple GitHub JobPermissions. The broader permission per use case will win
 *
 * @param permissions the permissions to merge
 * @return the merged permission
 */
export function mergeJobPermissions(...perms: JobPermissions[]): JobPermissions {
  const permissions: { [key: string]: JobPermission } = { ...perms[0] };

  for (const permission of perms.slice(1)) {
    for (const [key, value] of Object.entries(permission)) {
      permissions[key] = getBroadestPermission(value, permissions[key] ?? JobPermission.NONE);
    }
  }

  return permissions;
}

/**
 * Merge two GitHub JobPermission values. The broader permission will win
 *
 * @param perms the permissions to merge
 * @return the broadest permission
 */
export function getBroadestPermission(...perms: JobPermission[]): JobPermission {
  if (!perms || perms.length === 0) {
    throw new Error('No permissions provided');
  }
  for (const perm of perms) {
    if (!Object.values(JobPermission).includes(perm)) {
      throw new Error(`Invalid permission value: ${perm}`);
    }
  }

  if (perms.includes(JobPermission.WRITE)) {
    return JobPermission.WRITE;
  }
  if (perms.includes(JobPermission.READ)) {
    return JobPermission.READ;
  }
  return JobPermission.NONE;
}
