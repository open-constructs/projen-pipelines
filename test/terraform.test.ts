import { GitHubProject } from 'projen/lib/github';
import { synthSnapshot } from 'projen/lib/util/synth';
import { GithubTerraformPipeline, GitlabTerraformPipeline } from '../src';

test('Github snapshot', () => {
  const p = new GitHubProject({
    name: 'testapp',
  });

  new GithubTerraformPipeline(p, {
    iamRoleArns: {
      default: 'defaultRole',
    },
  });

  const snapshot = synthSnapshot(p);
  expect(snapshot['.github/workflows/deploy.yml']).toMatchSnapshot();
  expect(snapshot['.projen/tasks.json']).toMatchSnapshot();
});

test('Github snapshot with name', () => {
  const p = new GitHubProject({
    name: 'testapp',
  });

  new GithubTerraformPipeline(p, {
    name: 'foo',
    iamRoleArns: {
      default: 'defaultRole',
    },
  });

  const snapshot = synthSnapshot(p);
  expect(snapshot['.github/workflows/deploy-foo.yml']).toMatchSnapshot();
  expect(snapshot['.github/workflows/deploy.yml']).toBeUndefined();
  expect(snapshot['.projen/tasks.json']).toMatchSnapshot();
});

test('Gitlab snapshot', () => {
  const p = new GitHubProject({
    name: 'testapp',
  });

  new GitlabTerraformPipeline(p, {
    iamRoleArns: {
      default: 'defaultRole',
    },
  });

  const snapshot = synthSnapshot(p);
  expect(snapshot['.gitlab-ci.yml']).toMatchSnapshot();
});