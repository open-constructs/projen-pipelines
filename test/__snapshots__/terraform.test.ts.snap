// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Github snapshot 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

name: deploy
on:
  push:
    branches:
      - main
  workflow_dispatch: {}
jobs:
  plan:
    name: Plan Terraform changes
    runs-on: ubuntu-latest
    permissions:
      contents: read
      id-token: write
    env:
      CI: "true"
    steps:
      - name: Checkout
        uses: actions/checkout@v4
      - name: Setup Terraform
        uses: hashicorp/setup-terraform@v2
        with:
          terraform_version: latest
      - name: AWS Credentials
        uses: aws-actions/configure-aws-credentials@v4
        with:
          role-to-assume: defaultRole
          role-session-name: GitHubAction
          aws-region: us-east-1
      - run: npx projen tf:plan
  apply:
    name: Apply
    needs: plan
    runs-on: ubuntu-latest
    permissions:
      contents: read
      id-token: write
    env:
      CI: "true"
    steps:
      - name: Checkout
        uses: actions/checkout@v4
      - name: Setup Terraform
        uses: hashicorp/setup-terraform@v2
        with:
          terraform_version: latest
      - name: AWS Credentials
        uses: aws-actions/configure-aws-credentials@v4
        with:
          role-to-assume: defaultRole
          role-session-name: GitHubAction
          aws-region: us-east-1
      - run: npx projen tf:apply
"
`;

exports[`Github snapshot 2`] = `
{
  "//": "~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".",
  "tasks": {
    "build": {
      "description": "Full release build",
      "name": "build",
      "steps": [
        {
          "spawn": "default",
        },
        {
          "spawn": "pre-compile",
        },
        {
          "spawn": "compile",
        },
        {
          "spawn": "post-compile",
        },
        {
          "spawn": "test",
        },
        {
          "spawn": "package",
        },
      ],
    },
    "clobber": {
      "condition": "git diff --exit-code > /dev/null",
      "description": "hard resets to HEAD of origin and cleans the local repo",
      "env": {
        "BRANCH": "$(git branch --show-current)",
      },
      "name": "clobber",
      "steps": [
        {
          "exec": "git checkout -b scratch",
          "name": "save current HEAD in "scratch" branch",
        },
        {
          "exec": "git checkout $BRANCH",
        },
        {
          "exec": "git fetch origin",
          "name": "fetch latest changes from origin",
        },
        {
          "exec": "git reset --hard origin/$BRANCH",
          "name": "hard reset to origin commit",
        },
        {
          "exec": "git clean -fdx",
          "name": "clean all untracked files",
        },
        {
          "say": "ready to rock! (unpushed commits are under the "scratch" branch)",
        },
      ],
    },
    "compile": {
      "description": "Only compile",
      "name": "compile",
    },
    "default": {
      "description": "Synthesize project files",
      "name": "default",
    },
    "eject": {
      "description": "Remove projen from the project",
      "env": {
        "PROJEN_EJECTING": "true",
      },
      "name": "eject",
      "steps": [
        {
          "spawn": "default",
        },
      ],
    },
    "package": {
      "description": "Creates the distribution package",
      "name": "package",
    },
    "post-compile": {
      "description": "Runs after successful compilation",
      "name": "post-compile",
    },
    "pre-compile": {
      "description": "Prepare the project for compilation",
      "name": "pre-compile",
    },
    "test": {
      "description": "Run tests",
      "name": "test",
    },
    "tf:apply": {
      "description": "Apply terraform changes",
      "name": "tf:apply",
      "steps": [
        {
          "exec": "cd . && terraform init && terraform apply -auto-approve",
        },
      ],
    },
    "tf:destroy": {
      "description": "Destroy terraform resources",
      "name": "tf:destroy",
      "steps": [
        {
          "exec": "cd . && terraform init && terraform destroy -auto-approve",
        },
      ],
    },
    "tf:init": {
      "description": "Initialize terraform",
      "name": "tf:init",
      "steps": [
        {
          "exec": "cd . && terraform init",
        },
      ],
    },
    "tf:plan": {
      "description": "Run terraform plan",
      "name": "tf:plan",
      "steps": [
        {
          "exec": "cd . && terraform init && terraform plan",
        },
      ],
    },
    "tf:validate": {
      "description": "Validate terraform configuration",
      "name": "tf:validate",
      "steps": [
        {
          "exec": "cd . && terraform validate",
        },
      ],
    },
  },
}
`;

exports[`Github snapshot with name 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

name: deploy-foo
on:
  push:
    branches:
      - main
  workflow_dispatch: {}
jobs:
  plan:
    name: Plan Terraform changes
    runs-on: ubuntu-latest
    permissions:
      contents: read
      id-token: write
    env:
      CI: "true"
    steps:
      - name: Checkout
        uses: actions/checkout@v4
      - name: Setup Terraform
        uses: hashicorp/setup-terraform@v2
        with:
          terraform_version: latest
      - name: AWS Credentials
        uses: aws-actions/configure-aws-credentials@v4
        with:
          role-to-assume: defaultRole
          role-session-name: GitHubAction
          aws-region: us-east-1
      - run: npx projen tf:foo:plan
  apply:
    name: Apply
    needs: plan
    runs-on: ubuntu-latest
    permissions:
      contents: read
      id-token: write
    env:
      CI: "true"
    steps:
      - name: Checkout
        uses: actions/checkout@v4
      - name: Setup Terraform
        uses: hashicorp/setup-terraform@v2
        with:
          terraform_version: latest
      - name: AWS Credentials
        uses: aws-actions/configure-aws-credentials@v4
        with:
          role-to-assume: defaultRole
          role-session-name: GitHubAction
          aws-region: us-east-1
      - run: npx projen tf:foo:apply
"
`;

exports[`Github snapshot with name 2`] = `
{
  "//": "~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".",
  "tasks": {
    "build": {
      "description": "Full release build",
      "name": "build",
      "steps": [
        {
          "spawn": "default",
        },
        {
          "spawn": "pre-compile",
        },
        {
          "spawn": "compile",
        },
        {
          "spawn": "post-compile",
        },
        {
          "spawn": "test",
        },
        {
          "spawn": "package",
        },
      ],
    },
    "clobber": {
      "condition": "git diff --exit-code > /dev/null",
      "description": "hard resets to HEAD of origin and cleans the local repo",
      "env": {
        "BRANCH": "$(git branch --show-current)",
      },
      "name": "clobber",
      "steps": [
        {
          "exec": "git checkout -b scratch",
          "name": "save current HEAD in "scratch" branch",
        },
        {
          "exec": "git checkout $BRANCH",
        },
        {
          "exec": "git fetch origin",
          "name": "fetch latest changes from origin",
        },
        {
          "exec": "git reset --hard origin/$BRANCH",
          "name": "hard reset to origin commit",
        },
        {
          "exec": "git clean -fdx",
          "name": "clean all untracked files",
        },
        {
          "say": "ready to rock! (unpushed commits are under the "scratch" branch)",
        },
      ],
    },
    "compile": {
      "description": "Only compile",
      "name": "compile",
    },
    "default": {
      "description": "Synthesize project files",
      "name": "default",
    },
    "eject": {
      "description": "Remove projen from the project",
      "env": {
        "PROJEN_EJECTING": "true",
      },
      "name": "eject",
      "steps": [
        {
          "spawn": "default",
        },
      ],
    },
    "package": {
      "description": "Creates the distribution package",
      "name": "package",
    },
    "post-compile": {
      "description": "Runs after successful compilation",
      "name": "post-compile",
    },
    "pre-compile": {
      "description": "Prepare the project for compilation",
      "name": "pre-compile",
    },
    "test": {
      "description": "Run tests",
      "name": "test",
    },
    "tf:foo:apply": {
      "description": "Apply terraform changes",
      "name": "tf:foo:apply",
      "steps": [
        {
          "exec": "cd . && terraform init && terraform apply -auto-approve",
        },
      ],
    },
    "tf:foo:destroy": {
      "description": "Destroy terraform resources",
      "name": "tf:foo:destroy",
      "steps": [
        {
          "exec": "cd . && terraform init && terraform destroy -auto-approve",
        },
      ],
    },
    "tf:foo:init": {
      "description": "Initialize terraform",
      "name": "tf:foo:init",
      "steps": [
        {
          "exec": "cd . && terraform init",
        },
      ],
    },
    "tf:foo:plan": {
      "description": "Run terraform plan",
      "name": "tf:foo:plan",
      "steps": [
        {
          "exec": "cd . && terraform init && terraform plan",
        },
      ],
    },
    "tf:foo:validate": {
      "description": "Validate terraform configuration",
      "name": "tf:foo:validate",
      "steps": [
        {
          "exec": "cd . && terraform validate",
        },
      ],
    },
  },
}
`;

exports[`Gitlab snapshot 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

stages:
  - plan
  - apply
plan:
  extends: []
  needs: []
  stage: plan
  script:
    - export $(printf "AWS_ACCESS_KEY_ID=%s AWS_SECRET_ACCESS_KEY=%s AWS_SESSION_TOKEN=%s" $(aws sts assume-role-with-web-identity --role-arn "defaultRole" --role-session-name "GitLabRunner-\${CI_PROJECT_ID}-\${CI_PIPELINE_ID}}" --web-identity-token \${AWS_TOKEN} --duration-seconds 3600 --query 'Credentials.[AccessKeyId,SecretAccessKey,SessionToken]' --output text))
    - npx projen tf:plan
  variables: {}
  only:
    refs:
      - main
apply:
  extends: []
  needs: []
  stage: apply
  script:
    - export $(printf "AWS_ACCESS_KEY_ID=%s AWS_SECRET_ACCESS_KEY=%s AWS_SESSION_TOKEN=%s" $(aws sts assume-role-with-web-identity --role-arn "defaultRole" --role-session-name "GitLabRunner-\${CI_PROJECT_ID}-\${CI_PIPELINE_ID}}" --web-identity-token \${AWS_TOKEN} --duration-seconds 3600 --query 'Credentials.[AccessKeyId,SecretAccessKey,SessionToken]' --output text))
    - npx projen tf:apply
  variables: {}
  only:
    refs:
      - main
"
`;
